core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala

```diff
@@ -189,16 +199,68 @@ object ConsumerGroupCommand extends Logging {
     }
 
     def listGroups(): Unit = {
-      if (opts.options.has(opts.stateOpt)) {
-        val stateValue = opts.options.valueOf(opts.stateOpt)
-        val states = if (stateValue == null || stateValue.isEmpty)
-          Set[ConsumerGroupState]()
-        else
-          consumerGroupStatesFromString(stateValue)
-        val listings = listConsumerGroupsWithState(states)
-        printGroupStates(listings.map(e => (e.groupId, e.state.get.toString)))
-      } else
+      val includeType = opts.options.has(opts.typeOpt)
+      val includeState = opts.options.has(opts.stateOpt)
+
+      val groupInfoMap = mutable.Map[String, (String, String)]()
+
+      if (includeType || includeState) {
+        val types = getTypeValues()
+        val states = getStateValues()
+        val listings = {
+          listConsumerGroupsWithFilters(types, states)
+        }
+
+        listings.foreach { listing =>
+          val groupId = listing.groupId
+          val groupType = listing.groupType().orElse(GroupType.UNKNOWN).toString
+          val state = listing.state().orElse(ConsumerGroupState.UNKNOWN).toString
+          groupInfoMap.update(groupId, (groupType, state))
+        }
+
+        printGroupInfo(groupInfoMap, includeType, includeState)
+
+      } else {
         listConsumerGroups().foreach(println(_))
+      }
+    }
+
+    private def getStateValues(): Set[ConsumerGroupState] = {
+      val stateValue = opts.options.valueOf(opts.stateOpt)
+      if (stateValue == null || stateValue.isEmpty)
+        Set[ConsumerGroupState]()
+      else
+        consumerGroupStatesFromString(stateValue)
+    }
+
+    private def getTypeValues(): Set[GroupType] = {
+      val typeValue = opts.options.valueOf(opts.typeOpt)
+      if (typeValue == null || typeValue.isEmpty)
+        Set[GroupType]()
+      else
+        consumerGroupTypesFromString(typeValue)
+    }
+
+    private def printGroupInfo(groupsAndInfo: Map[String, (String, String)], includeType: Boolean, includeState: Boolean): Unit = {
+      val maxGroupLen: Int = groupsAndInfo.keys.foldLeft(15)((maxLen, groupId) => Math.max(maxLen, groupId.length))
+      var header = "GROUP"
+      var format = s"%-${maxGroupLen}s"
+
+      if (includeType) {
+        header += " TYPE"
+        format += " %-20s"
+      }
+      if (includeState) {
+        header += " STATE"
+        format += " %-20s"
+      }
+
+      println(format.format(ArraySeq.unsafeWrapArray(header.split(" ")): _*))
+
+      groupsAndInfo.foreach { case (groupId, (groupType, state)) =>
+        val info = List(groupId) ++ (if (includeType) List(groupType) else List()) ++ (if (includeState) List(state) else List())
```

Comment:
I was wondering we could simplify this a bit. One option would be to do the following:

```
    private def printGroupInfo(groups: List[ConsumerGroupListing], includeType: Boolean, includeState: Boolean): Unit = {
      def groupId(groupListing: ConsumerGroupListing): String = groupListing.groupId
      def groupType(groupListing: ConsumerGroupListing): String = groupListing.groupType.orElse(GroupType.UNKNOWN).toString
      def groupState(groupListing: ConsumerGroupListing): String = groupListing.state.orElse(ConsumerGroupState.UNKNOWN).toString

      val maxGroupLen = groups.foldLeft(15)((maxLen, groupListing) => Math.max(maxLen, groupId(groupListing).length))
      var format = s"%-${maxGroupLen}s"
      var header = List("Group")
      var extractors: List[ConsumerGroupListing => String] = List(groupId)

      if (includeType) {
        header += "TYPE"
        extractors += groupType
        format += " %-20s"
      }

      if (includeState) {
        header += "STATE"
        extractors += groupState
        format += " %-20s"
      }

      println(format.format(header: _*))

      groups.foreach { groupListing =>
        val info = extractors.map(extractor => extractor(groupListing))
        println(format.format(info: _*))
      }
    }
```

We could also think about making it fully generic in the future (not in this PR). For instance, we could have one function which takes a list of T, a list of header and a list of extractors. With this, you have everything to print the table.

Reply:
oo makes sense! Thanks!